radix-common/src/data/manifest/model/manifest_resource_assertion.rs

```diff
@@ -144,6 +148,46 @@ impl ManifestResourceConstraints {
     pub fn contains_specified_resource(&self, resource_address: &ResourceAddress) -> bool {
         self.specified_resources.contains_key(resource_address)
     }
+
+    pub fn validate(
+        self,
+        mut fungible_resources: BTreeMap<ResourceAddress, Decimal>,
+        mut non_fungible_resources: BTreeMap<ResourceAddress, IndexSet<NonFungibleLocalId>>,
+        exact: bool,
+    ) -> Result<(), ManifestResourceConstraintsError> {
+        for (resource_address, constraint) in self.specified_resources {
+            if resource_address.is_fungible() {
+                let amount = fungible_resources
+                    .remove(&resource_address)
+                    .unwrap_or_default();
+                constraint
+                    .validate_fungible(amount)
+                    .map_err(ManifestResourceConstraintsError::ResourceConstraint)?;
+            } else {
+                let ids = non_fungible_resources
+                    .remove(&resource_address)
+                    .unwrap_or_default()
+                    .clone();
+                constraint
+                    .validate_non_fungible(ids)
+                    .map_err(ManifestResourceConstraintsError::ResourceConstraint)?;
+            }
+        }
+
+        if exact {
+            if !fungible_resources.is_empty() || !non_fungible_resources.is_empty() {
```

Comment:
This is only correct if these are all non-zero balances. If this is required, it should IMO be expressed in the parameter name or rust doc. Currently I think this invariant _is_ correctly upheld inside the instruction processor, but it's a little indirect for my liking.

Instead, perhaps we could just iterate across the remaining `fungible_resources` and `non_fungible_resources` and if any of the balances are non-zero, then return an error with an example resource? That would then remove this pre-condition; and also allow us to give an example resource which was present.

Reply:
And I think we should add a test for this (to avoid potential regressions in future)

e.g.
* `ASSERT_NEXT_CALL_RETURNS_EXACTLY <empty>` and then `withdraw XRD Decimal("0")` should succeed
* `ASSERT_NEXT_CALL_RETURNS_EXACTLY <empty>` and then `withdraw <nfs> <empty>` should succeed

Reply:
Also, for good regression coverage, would be nice to have a quick test which applies an unmeetable `ASSERT_NEXT_CALL_RETURNS_EXACTLY` before a `CALL_FUNCTION` and both `YIELD_` commands (and e.g. check it gives a failure)

Reply:
> Also, for good regression coverage, would be nice to have a quick test which applies an unmeetable `ASSERT_NEXT_CALL_RETURNS_EXACTLY` before a `CALL_FUNCTION` and both `YIELD_` commands (and e.g. check it gives a failure)

This is already covered by tests `when_less_is_returned_assert_next_call_returns_only_should_fail` and `when_more_is_returned_assert_next_call_returns_only_should_fail` though it's missing the `YIELD_TO_CHILD` command. Will add this.